torchrl/data/utils.py

```diff
@@ -228,14 +228,15 @@ def _check_only_one_entry(
     recursive: bool = True,
 ):
     found_entry = False
-    for value in spec.values():
+    for key, value in spec.items():
         if isinstance(value, CompositeSpec):
             _check_only_one_entry(value, error, recursive)
         else:
-            if not found_entry:
-                found_entry = True
-            else:
-                raise error
+            if key == "done":
+                if found_entry:
+                    raise error
+                else:
+                    found_entry = True
```

Comment:
@matteobettini I'm not even sure we need this function anymore. We wanted to avoid that people would have `"done_agent1"` and `"done_agent2"` and rather `("agent1", "done")` IIRC. 
Now we will allow `"done_spec"` to contain `"done"` and `"truncated"` at any level. Per se the `done_spec` could even be richer. So I don't think we need this check anymore: it's good practice to have only one done per level but I don't think that will break anywhere if you have more.

The only exception I see is computing the `"_reset"` signal which usually reflects `done | truncated`. In other words, `_reset` reflects somethings like
```
def _make_reset(data, done_spec):
    out = {}
    for key, item in done_spec.items():
        if isinstance(done_spec, CompositeSpec):
            out[key] = _make_reset(data.get(key), item)
        else:
            out["_reset"] = out.get("_reset", False) | data.get(key)
     return data.update(out)
```
This means tht if you put done_agent1 and done_agent2 at the same level you'll get a single reset for both. Not ideal but not really our business if the MARL API is well documented IMO.

In collectors we use `"done"` to compute the trajectory ids but this is an extra feature that could apply only in settings where done is at the root (or in other restricted settings). The trajectory id doesn't influence the collector behaviour, which only cares about the number of frames (number of elements of the root tensordict). 

For the rest, loss modules have parametric dones so I think we could rename `"done"` to `"merylstreep"` and they could still work. They will need refactoring to account for `"truncated"` anyway.

Reply:
So the situation is a little more complex, let me try to explain it here:

Since we decided that our naming convention is to have one "_reset" key per tensordict (due to the fact that if we had more we would have to change the name), then this key has to reflect the shape of  the "done"  in that tensordict.

If we allowed more then one leaf key in each td of the done_spec then the problem is what shape should the _reset reflect. We cannot do like vincent said cause the shapes might be arbitrary.

Therefore the convention established is that at least a key with the name "done" (and exactly done) should exist in each td of the done_spec and the reset will mimic its shape.

If now we want to change the convention and say that the constraint now is that all the values in the same tensordict of the done_spec should have the same shape, that is fine by me, but then we have to check and enforce that.

Reply:
if we do like in this PR and assume to always have truncated in the done_spec, then we have to carefully adjust all the components of the codebase that create the _reset key.

Like here
https://github.com/pytorch/rl/blob/fe91c4fc0e7ec6bde969a4f25ea58b247ec4dc65/torchrl/collectors/collectors.py#L797

and here
https://github.com/pytorch/rl/blob/fe91c4fc0e7ec6bde969a4f25ea58b247ec4dc65/torchrl/envs/common.py#L1590

For example.

These line will currently create a `_reset` for each `done_spec.keys(True,True)`.
But if the `done_spec` now contains more than one leaf per td we have to change all this.

Reply:
> If now we want to change the convention and say that the constraint now is that all the values in the same tensordict of the done_spec should have the same shape, that is fine by me, but then we have to check and enforce that.

Per se, if someone writes an env that does not require `_reset` (which we use internally but is not exposed to users) it's fine. What I mean is that if a data structure has `"done_agent1"` of shape `X` and `"done_agent2"` of shape `Y`, computing a `"_reset"` key will fail but that only means that the env cannot be used within a collector.
Since we want to provide independent, modular components, I think it's a design choice that is ok not to enforce that all envs written by any user would fit in a collector. We need to make sure that the collector tells you specifically what is wrong with your env though.

What about letting envs have the `done_spec` they want, and let the collectors / parallel envs spot the potential problems during construction?

Reply:
Yeah but also the rollout function uses the reset logic. So how would we treat that?

Reply:
Imagine someone wants to use TorchRL to write a multiagent env that does some job with the `"done_agentX"` structure.
I think it's cool if they do that and put that in their app. I don't think they should get an exception or a warning that tells them that they're going to be in a dead end when using a collector. Maybe they don't want to use the collectors.
Let's not prevent legitimate use of one class to prevent bug in another if the usage of the other is optional, wdyt?

Reply:
Yes but vectorized envs will need to be fed a `_reset` flag upon reset in the `env.rollout` function.  so you are saying we will provide that with our best effort but if we cannot we just give a warning and say that in collector this will be an error?

Reply:
> Yeah but also the rollout function uses the reset logic. So how would we treat that?

Oh yeah good point.
This is a bit tricky: per se, we set a `_reset` in rollout and hope for the best but we don't enforce that it's being read anywhere.
The thing is that we don't really have any agency about what `reset(tensordict)` does under the hood. For ParallelEnv and similar we know what to do but in other cases I'm not 100% sure of what could be done generically.

But yeah, since we use `_reset` within the env class it makes sense that it's computed in a consistent way within the library. 
So what we would opt for is:
- during construction, walk through the specs and check that at each level, all done specs have a matching shape
- during execution, compute `_reset` as the combination of all `_reset` keys at each level

Reply:
> during execution, compute _reset as the combination of all _reset keys at each level

->  as the combination of all **done_spec_keys** at each level

but yes that could work!